modules/core/src/main/java/org/apache/ignite/internal/processors/cache/tree/updatelog/PartitionLogTree.java

```diff
@@ -141,4 +166,78 @@ public PartitionLogTree(
 
         return flag == FULL_ROW ? row.initRow(grp) : row;
     }
+
+    /**
+     * Construct the exception and invoke failure processor.
+     *
+     * @param msg Message.
+     * @param cause Cause.
+     * @param grpId Group id.
+     * @param pageIds Pages ids.
+     * @return New CorruptedTreeException instance.
+     */
+    @Override protected CorruptedTreeException corruptedTreeException(String msg, Throwable cause, int grpId,
+        long... pageIds) {
+        if (cause instanceof DuplicateUpdateCounterException) {
+            // Duplicate update counter doesn't mean the tree is corrupted.
+            return super.corruptedTreeException(msg, cause, grpId, pageIds);
+        }
+
+        CorruptedTreeException e = new CorruptedTreeException(msg, cause, grpName, null, name(), grpId, pageIds);
+
+        String errorMsg = "Partition log tree of partition `" + part + "` of cache group `" + grpName + "` is corrupted, " +
+            "to fix this issue a rebuild is required. On the next restart, node will enter " +
+            "the maintenance mode and rebuild corrupted tree.";
+
+        log.warning(errorMsg);
+
+        try {
+            MaintenanceTask task = toMaintenanceTask(grpId);
+
+            grp.shared().kernalContext().maintenanceRegistry().registerMaintenanceTask(task,
+                oldTask -> mergeTasks(oldTask, task));
+        }
+        catch (IgniteCheckedException ex) {
+            log.warning("Failed to register maintenance record for corrupted partition files.", ex);
+        }
+
+        processFailure(FailureType.CRITICAL_ERROR, e);
+
+        return e;
+    }
+
+    /**
+     * Constructs a partition log tree rebuild maintenance task.
+     *
+     * @param groupId Group id.
+     * @return Maintenance task.
+     */
+    public static MaintenanceTask toMaintenanceTask(int groupId) {
+        return new MaintenanceTask(PART_LOG_TREE_REBUILD_MNTC_TASK_NAME, TASK_DESCRIPTION,
+            U.hexInt(groupId));
```

Comment:
Because we use String.contains() to check if the cache group id is in the list.
This may cause false negative result for e.g. next groups ids 25 and 5.

Reply:
Ok